Nested type conformance problem

With swift-driver version: 1.45.2 Apple Swift version 5.6.1 (swiftlang-5.6.0.323.66 clang-1316.0.20.12)
Target: arm64-apple-macosx12.0

```swift
enum S0: Int, Comparable { // compiles fine
  static func < (lhs: Self, rhs: Self) -> Bool { lhs.rawValue < rhs.rawValue }
  case A, B
}

struct Test {
  func test() {
    enum S1: Int, Comparable { // fails
      static func < (lhs: Self, rhs: Self) -> Bool { lhs.rawValue < rhs.rawValue }
      case A, B
    }
  }
}
```

should compile, but yields:

```
x.swift:8:10: error: type 'S1' does not conform to protocol 'Comparable'
    enum S1: Int, Comparable {
         ^
x.swift:8:14: note: enum declares raw type 'Int', preventing synthesized conformance of 'S1' to 'Comparable'
    enum S1: Int, Comparable {
             ^
Swift.FloatingPoint:3:24: note: candidate would match if 'S1' conformed to 'FloatingPoint'
    public static func < (lhs: Self, rhs: Self) -> Bool
                       ^
Swift.BinaryInteger:4:24: note: candidate has non-matching type '<Self, Other> (Self, Other) -> Bool'
    public static func < <Other>(lhs: Self, rhs: Other) -> Bool where Other : BinaryInteger
                       ^
Swift._Pointer:2:24: note: candidate would match if 'S1' conformed to '_Pointer'
    public static func < (lhs: Self, rhs: Self) -> Bool
                       ^
Swift.Strideable:2:35: note: candidate has non-matching type '<Self> (Self, Self) -> Bool'
    @inlinable public static func < (x: Self, y: Self) -> Bool
                                  ^
Swift.StringProtocol:4:35: note: candidate would match if 'S1' conformed to 'StringProtocol'
    @inlinable public static func < <RHS>(lhs: Self, rhs: RHS) -> Bool where RHS : StringProtocol
                                  ^
Swift.Comparable:2:17: note: protocol requires function '<' with type '(S1, S1) -> Bool'
    static func < (lhs: Self, rhs: Self) -> Bool
                ^
```

--- a/lib/Sema/TypeCheckProtocol.cpp
+++ b/lib/Sema/TypeCheckProtocol.cpp
@@ -1315,25 +1315,41 @@ WitnessChecker::lookupValueWitnesses(ValueDecl *req, bool *ignoringNames) {
   auto reqName = req->createNameRef();
   auto reqBaseName = reqName.withoutArgumentLabels();
 
-  if (req->isOperator()) {
-    // Operator lookup is always global.
+  // An operator function is the only kind of witness that requires global
+  // lookup. However, because global lookup doesn't enter local contexts,
+  // an additional, qualified lookup is warranted when the conforming type
+  // is declared in a local context.
+  const bool doUnqualifiedLookup = req->isOperator();
+  const bool doQualifiedLookup =
+      !req->isOperator() || DC->getParent()->getLocalContext();
+
+  if (doUnqualifiedLookup) {
     auto lookup = TypeChecker::lookupUnqualified(DC->getModuleScopeContext(),
                                                  reqBaseName, SourceLoc(),
                                                  defaultUnqualifiedLookupOptions);
     for (auto candidate : lookup) {
       auto decl = candidate.getValueDecl();
-      if (swift::isMemberOperator(cast<FuncDecl>(decl), Adoptee)) {
+      if (!isa<ProtocolDecl>(decl->getDeclContext()) &&
+          swift::isMemberOperator(cast<FuncDecl>(decl), Adoptee)) {
         witnesses.push_back(decl);
       }
     }
-  } else {
-    // Variable/function/subscript requirements.
+  }
+
+  if (doQualifiedLookup) {
     auto *nominal = Adoptee->getAnyNominal();
     nominal->synthesizeSemanticMembersIfNeeded(reqName.getFullName());
 
+    // Unqualified lookup would have already found candidates from protocol
+    // extensions, including those that match only by base name. Take care not
+    // to restate them in the resulting list, or else an otherwise valid
+    // conformance will become ambiguous.
+    const NLOptions options =
+        doUnqualifiedLookup ? NLOptions(0) : NL_ProtocolMembers;
+
     SmallVector<ValueDecl *, 4> lookupResults;
     bool addedAny = false;
-    DC->lookupQualified(nominal, reqName, NL_ProtocolMembers, lookupResults);
+    DC->lookupQualified(nominal, reqName, options, lookupResults);
     for (auto *decl : lookupResults) {
       if (!isa<ProtocolDecl>(decl->getDeclContext())) {
         witnesses.push_back(decl);
@@ -1345,7 +1361,7 @@ WitnessChecker::lookupValueWitnesses(ValueDecl *req, bool *ignoringNames) {
     // again using only the base name.
     if (!addedAny && ignoringNames) {
       lookupResults.clear();
-      DC->lookupQualified(nominal, reqBaseName, NL_ProtocolMembers, lookupResults);
+      DC->lookupQualified(nominal, reqBaseName, options, lookupResults);
       for (auto *decl : lookupResults) {
         if (!isa<ProtocolDecl>(decl->getDeclContext()))
           witnesses.push_back(decl);
@@ -1357,6 +1373,10 @@ WitnessChecker::lookupValueWitnesses(ValueDecl *req, bool *ignoringNames) {
     removeOverriddenDecls(witnesses);
     removeShadowedDecls(witnesses, DC);
   }
+
+  assert(llvm::none_of(witnesses, [](ValueDecl *decl) {
+    return isa<ProtocolDecl>(decl->getDeclContext());
+  }));
 
   return witnesses;
 }

--- a/test/SILGen/protocol_operators_local_conformance.swift
+++ b/test/SILGen/protocol_operators_local_conformance.swift
@@ -1,0 +1,87 @@
+// RUN: %target-swift-frontend -emit-silgen %s | %FileCheck %s
+//
+// Test that we find and pick the right witnesses for operator function
+// requirements when the conforming type is declared in a local context.
+
+infix operator =*=
+protocol P1 {
+  static func =*= (lhs: Self, rhs: Self) -> Bool
+}
+func test1() {
+  struct S: P1 {
+    static func =*= (lhs: S, rhs: S) -> Bool { true }
+  }
+}
+// CHECK-LABEL: sil private [transparent] [thunk] [ossa] @$s36protocol_operators_local_conformance5test1yyF1SL_VAA2P1A2aEP3emeoiySbx_xtFZTW : $@convention(witness_method: P1) (@in_guaranteed S, @in_guaranteed S, @thick S.Type) -> Bool {
+// CHECK: function_ref @$s36protocol_operators_local_conformance5test1yyF1SL_V3emeoiySbAD_ADtFZ : $@convention(method) (S, S, @thin S.Type) -> Bool
+
+infix operator =**=
+protocol P2 {
+  static func =**= (lhs: Self, rhs: Self) -> Bool
+}
+extension P2 {
+  static func =**= (lhs: Self, rhs: Self) -> Bool { true }
+}
+func test2() {
+  struct S: P2 {}
+}
+// CHECK-LABEL: sil private [transparent] [thunk] [ossa] @$s36protocol_operators_local_conformance5test2yyF1SL_VAA2P2A2aEP4emmeoiySbx_xtFZTW : $@convention(witness_method: P2) (@in_guaranteed S, @in_guaranteed S, @thick S.Type) -> Bool {
+// CHECK: function_ref @$s36protocol_operators_local_conformance2P2PAAE4emmeoiySbx_xtFZ : $@convention(method) <τ_0_0 where τ_0_0 : P2> (@in_guaranteed τ_0_0, @in_guaranteed τ_0_0, @thick τ_0_0.Type) -> Bool
+
+func test3() {
+  struct S: P2 {
+    static func =**= (lhs: S, rhs: S) -> Bool { true }
+  }
+}
+// CHECK-LABEL: sil private [transparent] [thunk] [ossa] @$s36protocol_operators_local_conformance5test3yyF1SL_VAA2P2A2aEP4emmeoiySbx_xtFZTW : $@convention(witness_method: P2) (@in_guaranteed S, @in_guaranteed S, @thick S.Type) -> Bool {
+// CHECK: function_ref @$s36protocol_operators_local_conformance5test3yyF1SL_V4emmeoiySbAD_ADtFZ : $@convention(method) (S, S, @thin S.Type) -> Bool
+
+func test4() {
+  enum E: Int, Equatable {
+    case a, b
+
+    static func == (lhs: E, rhs: E) -> Bool { true }
+  }
+}
+// CHECK-LABEL: sil private [transparent] [thunk] [ossa] @$s36protocol_operators_local_conformance5test4yyF1EL_OSQAASQ2eeoiySbx_xtFZTW : $@convention(witness_method: Equatable) (@in_guaranteed E, @in_guaranteed E, @thick E.Type) -> Bool {
+// CHECK: function_ref @$s36protocol_operators_local_conformance5test4yyF1EL_O2eeoiySbAD_ADtFZ : $@convention(method) (E, E, @thin E.Type) -> Bool
+
+func test5() {
+  enum E: Equatable {
+    case a, b
+  }
+}
+// CHECK-LABEL: sil private [transparent] [thunk] [ossa] @$s36protocol_operators_local_conformance5test5yyF1EL_OSQAASQ2eeoiySbx_xtFZTW : $@convention(witness_method: Equatable) (@in_guaranteed E, @in_guaranteed E, @thick E.Type) -> Bool {
+// CHECK: function_ref @$s36protocol_operators_local_conformance5test5yyF1EL_O21__derived_enum_equalsySbAD_ADtFZ : $@convention(method) (E, E, @thin E.Type) -> Bool
+
+
+func test6() {
+  enum E: Int, Equatable {
+    case a, b
+  }
+}
+// CHECK-LABEL: sil private [transparent] [thunk] [ossa] @$s36protocol_operators_local_conformance5test6yyF1EL_OSQAASQ2eeoiySbx_xtFZTW : $@convention(witness_method: Equatable) (@in_guaranteed E, @in_guaranteed E, @thick E.Type) -> Bool {
+// CHECK: function_ref @$[[TEST6_EQUALS_WITNESS:[_0-9a-zA-Z]+]]
+// CHECK: }
+
+// CHECK: sil [serialized] @$[[TEST6_EQUALS_WITNESS]] : $@convention(thin) <τ_0_0 where τ_0_0 : RawRepresentable, τ_0_0.RawValue : Equatable> (@in_guaranteed τ_0_0, @in_guaranteed τ_0_0) -> Bool
+
+func test7() {
+  struct Outer {
+    enum Inner: Int, Comparable {
+      case a, b
+
+      static func < (lhs: Inner, rhs: Inner) -> Bool { true }
+    }
+  }
+}
+// CHECK-LABEL: sil private [transparent] [thunk] [ossa] @$s36protocol_operators_local_conformance5test7yyF5OuterL_V5InnerOSLAASL1loiySbx_xtFZTW : $@convention(witness_method: Comparable) (@in_guaranteed Outer.Inner, @in_guaranteed Outer.Inner, @thick Outer.Inner.Type) -> Bool {
+// CHECK: function_ref @$s36protocol_operators_local_conformance5test7yyF5OuterL_V5InnerO1loiySbAF_AFtFZ : $@convention(method) (Outer.Inner, Outer.Inner, @thin Outer.Inner.Type) -> Bool
+
+func test8() {
+  class C: Equatable {
+    static func == (lhs: C, rhs: C) -> Bool { true }
+  }
+}
+// CHECK-LABEL: sil private [transparent] [thunk] [ossa] @$s36protocol_operators_local_conformance5test8yyF1CL_CSQAASQ2eeoiySbx_xtFZTW : $@convention(witness_method: Equatable) (@in_guaranteed C, @in_guaranteed C, @thick C.Type) -> Bool {
+// CHECK: function_ref @$s36protocol_operators_local_conformance5test8yyF1CL_C2eeoiySbAD_ADtFZ : $@convention(method) (@guaranteed C, @guaranteed C, @thick C.Type) -> Bool

--- a/test/decl/protocol/conforms/operator_local_conformance.swift
+++ b/test/decl/protocol/conforms/operator_local_conformance.swift
@@ -1,0 +1,46 @@
+// RUN: %target-typecheck-verify-swift
+
+infix operator =*=
+protocol P1 {
+  static func =*= (lhs: Self, rhs: Self) -> Bool
+}
+extension P1 {
+  static func =*= (lhs: Self, rhs: Self) -> Bool { true }
+}
+func test1() {
+  struct S: P1 {}
+}
+
+infix operator =**=
+protocol P2 {
+  static func =**= (lhs: Self, rhs: Self) -> Bool
+}
+func test2() {
+  struct S: P2 {
+    static func =**= (lhs: S, rhs: S) -> Bool { true }
+  }
+}
+
+func test3() {
+  enum E: Equatable {
+    case a, b
+
+    static func == (lhs: E, rhs: E) -> Bool { true }
+  }
+}
+
+func test4() {
+  struct Outer {
+    enum Inner: Int, Comparable {
+      case a, b
+
+      static func < (lhs: Inner, rhs: Inner) -> Bool { true }
+    }
+  }
+}
+
+func test5() {
+  class C: Equatable {
+    static func == (lhs: C, rhs: C) -> Bool { true }
+  }
+}

